htdocs/core/modules/modSociete.class.php

```diff
@@ -208,6 +208,14 @@ function __construct($db)
 		$this->rights[$r][3] = 0;
 		$this->rights[$r][4] = 'client';
 		$this->rights[$r][5] = 'voir';
+		
```

Comment:
This break the standard behaviour of Dolibarr that say :

-  "Permission in Dolibarr are read / write for all" except for:
- For external user limited to themself
- For RH module (expense report, leave request) where permission is ALWAYS a permission for all subordinates too.
- For project where permission are owned if we are assigned or not to project.


Introducing this will creates a ton of problem. What to do when a projet is allowed to a user but on a thirdparty not allowed to user, same for member, what if thirdparty is external user...
Also the getAllChildIds is a heavy function that will make all Dolibarr slowed seriously on large database.
And there is a ton of other side effect forcing to put a DISTINCT on all sql request everywhere (it sucks).
So i would prefer to have this managed by an external module (it will be a pain to manage this by core).

So this is what i suggest :
We can introduce into requests the  

..." IN (".getUserAndSub().")";

By default, it will return the user id so there is no side effect.
We also introduce a  getDistinctOrNot  function.

So external modules will be able to return or not the DISTINCT and return the list of restricted users.
All problems will be in external module so no side effects on Dolibarr core.